File: android/app/src/main/java/ca/tristan/portfolio/ui/format/TickerFlag.kt
package ca.tristan.portfolio.ui.format

import ca.tristan.portfolio.data.MarketIndices

/**
 * Country flag for a ticker/exchange pair, shared by every screen that lists
 * tickers (search results, watchlist rows, holding rows).
 *
 * [MarketScreen] and [QuoteDetailScreen] each grew their own private copy of
 * this before this file existed; those aren't touched here since they aren't
 * the ones with a reported bug, but any *new* row this round (search results,
 * the holding-detail header) reads from here so there's exactly one mapping
 * to keep current when a new exchange suffix shows up.
 */
object TickerFlag {

    /**
     * Best-effort flag for [ticker], preferring an explicit Yahoo-style
     * [exchange] label (from search results — "NASDAQ", "Toronto", "London",
     * an ISO exchange code, etc.) when the ticker's own suffix doesn't say
     * enough, then falling back to [currency].
     */
    fun forTicker(ticker: String, exchange: String? = null, currency: String? = null): String {
        val t = ticker.trim().uppercase()
        MarketIndices.flag(t).takeIf { it.isNotEmpty() }?.let { return it }
        if (t.startsWith("^")) return ""
        if (t.contains("-USD") || t.contains("-BTC") || t.contains("-ETH")) return ""

        bySuffix(t)?.let { return it }
        exchange?.let { byExchangeName(it) }?.let { return it }
        return byCurrency(currency) ?: "🇺🇸"
    }

    /**
     * Flag for a bare currency code, or "" when it isn't one this knows.
     *
     * Unlike [forTicker] there is no 🇺🇸 default here: the callers are
     * currency section headers, where guessing a flag for an unrecognised code
     * would label a bucket with the wrong country outright.
     */
    fun forCurrency(currency: String?): String = byCurrency(currency) ?: ""

    private fun bySuffix(t: String): String? = when {
        t.endsWith(".TO") || t.endsWith(".V") || t.endsWith(".NE") ||
            t.endsWith(".CN") || t.endsWith(".TSX")                -> "🇨🇦"
        t.endsWith(".L")                                           -> "🇬🇧"
        t.endsWith(".DE") || t.endsWith(".F") || t.endsWith(".BE") -> "🇩🇪"
        t.endsWith(".PA")                                          -> "🇫🇷"
        t.endsWith(".AS")                                          -> "🇳🇱"
        t.endsWith(".SW")                                          -> "🇨🇭"
        t.endsWith(".MI")                                          -> "🇮🇹"
        t.endsWith(".MC")                                          -> "🇪🇸"
        t.endsWith(".ST")                                          -> "🇸🇪"
        t.endsWith(".OL")                                          -> "🇳🇴"
        t.endsWith(".CO")                                          -> "🇩🇰"
        t.endsWith(".HE")                                          -> "🇫🇮"
        t.endsWith(".IC")                                          -> "🇮🇸"
        t.endsWith(".VI")                                          -> "🇦🇹"
        t.endsWith(".BR")                                          -> "🇧🇪"
        t.endsWith(".LS")                                          -> "🇵🇹"
        t.endsWith(".IR")                                          -> "🇮🇪"
        t.endsWith(".WA")                                          -> "🇵🇱"
        t.endsWith(".T")                                           -> "🇯🇵"
        t.endsWith(".HK")                                          -> "🇭🇰"
        t.endsWith(".SS") || t.endsWith(".SZ")                     -> "🇨🇳"
        t.endsWith(".AX")                                          -> "🇦🇺"
        t.endsWith(".NZ")                                          -> "🇳🇿"
        t.endsWith(".NS") || t.endsWith(".BO")                     -> "🇮🇳"
        t.endsWith(".SA")                                          -> "🇧🇷"
        t.endsWith(".MX")                                          -> "🇲🇽"
        t.endsWith(".KS") || t.endsWith(".KQ")                     -> "🇰🇷"
        t.endsWith(".TW") || t.endsWith(".TWO")                    -> "🇹🇼"
        t.endsWith(".SI")                                          -> "🇸🇬"
        t.endsWith(".JK")                                          -> "🇮🇩"
        t.endsWith(".PS")                                          -> "🇵🇭"
        t.endsWith(".KL")                                          -> "🇲🇾"
        t.endsWith(".BK")                                          -> "🇹🇭"
        t.endsWith(".IS")                                          -> "🇹🇷"
        t.endsWith(".TA")                                          -> "🇮🇱"
        t.endsWith(".JO")                                          -> "🇿🇦"
        else -> null
    }

    private fun byExchangeName(exchange: String): String? {
        val e = exchange.uppercase()
        return when {
            "TORONTO" in e || "TSX" in e || "VENTURE" in e || "NEO" in e || "CBOE CA" in e -> "🇨🇦"
            "LONDON" in e || e == "LSE"                                                     -> "🇬🇧"
            "FRANKFURT" in e || "XETRA" in e                                                -> "🇩🇪"
            "PARIS" in e || "EURONEXT" in e                                                 -> "🇫🇷"
            "STOCKHOLM" in e                                                                -> "🇸🇪"
            "OSLO" in e                                                                     -> "🇳🇴"
            "COPENHAGEN" in e                                                               -> "🇩🇰"
            "HELSINKI" in e                                                                 -> "🇫🇮"
            "TOKYO" in e || "OSAKA" in e                                                    -> "🇯🇵"
            "HONG KONG" in e                                                                -> "🇭🇰"
            "SHANGHAI" in e || "SHENZHEN" in e                                              -> "🇨🇳"
            "SYDNEY" in e || "ASX" in e                                                     -> "🇦🇺"
            "NSE" in e || "BSE" in e || "BOMBAY" in e                                       -> "🇮🇳"
            "SAO PAULO" in e || "BOVESPA" in e || "B3" in e                                 -> "🇧🇷"
            "MEXICO" in e                                                                   -> "🇲🇽"
            "KOREA" in e || "KOSPI" in e || "KOSDAQ" in e                                   -> "🇰🇷"
            "TAIWAN" in e                                                                   -> "🇹🇼"
            "SINGAPORE" in e                                                                -> "🇸🇬"
            "JAKARTA" in e || "INDONESIA" in e                                              -> "🇮🇩"
            "PHILIPPINE" in e || "MANILA" in e || e == "PSE"                                -> "🇵🇭"
            "KUALA LUMPUR" in e || "MALAYSIA" in e                                          -> "🇲🇾"
            "NASDAQ" in e || "NYSE" in e || "AMEX" in e || "NYQ" in e || "NMS" in e ||
                "PCX" in e || "BATS" in e                                                    -> "🇺🇸"
            else -> null
        }
    }

    private fun byCurrency(currency: String?): String? = when (currency?.uppercase()) {
        "CAD" -> "🇨🇦"; "GBP", "GBP.L", "GBX" -> "🇬🇧"; "EUR" -> "🇪🇺"
        "JPY" -> "🇯🇵"; "AUD" -> "🇦🇺"; "HKD" -> "🇭🇰"
        "CHF" -> "🇨🇭"; "INR" -> "🇮🇳"; "BRL" -> "🇧🇷"
        "SEK" -> "🇸🇪"; "NOK" -> "🇳🇴"; "DKK" -> "🇩🇰"
        "KRW" -> "🇰🇷"; "TWD" -> "🇹🇼"; "SGD" -> "🇸🇬"
        "MXN" -> "🇲🇽"; "CNY" -> "🇨🇳"; "NZD" -> "🇳🇿"
        "ISK" -> "🇮🇸"; "PLN" -> "🇵🇱"; "THB" -> "🇹🇭"
        "TRY" -> "🇹🇷"; "ILS" -> "🇮🇱"; "ZAR" -> "🇿🇦"
        "IDR" -> "🇮🇩"; "PHP" -> "🇵🇭"; "MYR" -> "🇲🇾"
        "USD" -> "🇺🇸"
        else -> null
    }
}

File: ios/WealthBoard/Format/TickerFlag.swift
import Foundation

/// Country flag for a ticker/exchange pair, shared by every screen that lists
/// tickers (search results, watchlist rows, holding rows).
///
/// Ported from `ui/format/TickerFlag.kt` — one mapping for the whole app, so a
/// new exchange suffix is added in exactly one place.
enum TickerFlag {

    /// Best-effort flag for `ticker`, preferring an explicit Yahoo-style
    /// `exchange` label when the ticker's own suffix doesn't say enough, then
    /// falling back to `currency`.
    static func forTicker(_ ticker: String, exchange: String? = nil, currency: String? = nil) -> String {
        let t = ticker.trimmingCharacters(in: .whitespaces).uppercased()

        let indexFlag = MarketIndices.flag(for: t)
        if !indexFlag.isEmpty { return indexFlag }

        if t.hasPrefix("^") { return "" }
        if t.contains("-USD") || t.contains("-BTC") || t.contains("-ETH") { return "" }

        if let s = bySuffix(t) { return s }
        if let e = exchange, let f = byExchangeName(e) { return f }
        return byCurrency(currency) ?? "🇺🇸"
    }

    /// Flag for a bare currency code, or "" when it isn't one this knows.
    ///
    /// Unlike `forTicker` there is no 🇺🇸 default here: the callers are currency
    /// section headers, where guessing a flag for an unrecognised code would
    /// label a bucket with the wrong country outright.
    static func forCurrency(_ currency: String?) -> String {
        byCurrency(currency) ?? ""
    }

    private static func bySuffix(_ t: String) -> String? {
        if t.hasSuffix(".TO") || t.hasSuffix(".V") || t.hasSuffix(".NE")
            || t.hasSuffix(".CN") || t.hasSuffix(".TSX") { return "🇨🇦" }
        if t.hasSuffix(".L") { return "🇬🇧" }
        if t.hasSuffix(".DE") || t.hasSuffix(".F") || t.hasSuffix(".BE") { return "🇩🇪" }
        if t.hasSuffix(".PA") { return "🇫🇷" }
        if t.hasSuffix(".AS") { return "🇳🇱" }
        if t.hasSuffix(".SW") { return "🇨🇭" }
        if t.hasSuffix(".MI") { return "🇮🇹" }
        if t.hasSuffix(".MC") { return "🇪🇸" }
        if t.hasSuffix(".ST") { return "🇸🇪" }
        if t.hasSuffix(".OL") { return "🇳🇴" }
        if t.hasSuffix(".CO") { return "🇩🇰" }
        if t.hasSuffix(".HE") { return "🇫🇮" }
        if t.hasSuffix(".IC") { return "🇮🇸" }
        if t.hasSuffix(".VI") { return "🇦🇹" }
        if t.hasSuffix(".BR") { return "🇧🇪" }
        if t.hasSuffix(".LS") { return "🇵🇹" }
        if t.hasSuffix(".IR") { return "🇮🇪" }
        if t.hasSuffix(".WA") { return "🇵🇱" }
        if t.hasSuffix(".T") { return "🇯🇵" }
        if t.hasSuffix(".HK") { return "🇭🇰" }
        if t.hasSuffix(".SS") || t.hasSuffix(".SZ") { return "🇨🇳" }
        if t.hasSuffix(".AX") { return "🇦🇺" }
        if t.hasSuffix(".NZ") { return "🇳🇿" }
        if t.hasSuffix(".NS") || t.hasSuffix(".BO") { return "🇮🇳" }
        if t.hasSuffix(".SA") { return "🇧🇷" }
        if t.hasSuffix(".MX") { return "🇲🇽" }
        if t.hasSuffix(".KS") || t.hasSuffix(".KQ") { return "🇰🇷" }
        if t.hasSuffix(".TW") || t.hasSuffix(".TWO") { return "🇹🇼" }
        if t.hasSuffix(".SI") { return "🇸🇬" }
        if t.hasSuffix(".JK") { return "🇮🇩" }
        if t.hasSuffix(".PS") { return "🇵🇭" }
        if t.hasSuffix(".KL") { return "🇲🇾" }
        if t.hasSuffix(".BK") { return "🇹🇭" }
        if t.hasSuffix(".IS") { return "🇹🇷" }
        if t.hasSuffix(".TA") { return "🇮🇱" }
        if t.hasSuffix(".JO") { return "🇿🇦" }
        return nil
    }

    private static func byExchangeName(_ exchange: String) -> String? {
        let e = exchange.uppercased()
        func has(_ needle: String) -> Bool { e.contains(needle) }

        if has("TORONTO") || has("TSX") || has("VENTURE") || has("NEO") || has("CBOE CA") { return "🇨🇦" }
        if has("LONDON") || e == "LSE" { return "🇬🇧" }
        if has("FRANKFURT") || has("XETRA") { return "🇩🇪" }
        if has("PARIS") || has("EURONEXT") { return "🇫🇷" }
        if has("STOCKHOLM") { return "🇸🇪" }
        if has("OSLO") { return "🇳🇴" }
        if has("COPENHAGEN") { return "🇩🇰" }
        if has("HELSINKI") { return "🇫🇮" }
        if has("TOKYO") || has("OSAKA") { return "🇯🇵" }
        if has("HONG KONG") { return "🇭🇰" }
        if has("SHANGHAI") || has("SHENZHEN") { return "🇨🇳" }
        if has("SYDNEY") || has("ASX") { return "🇦🇺" }
        if has("NSE") || has("BSE") || has("BOMBAY") { return "🇮🇳" }
        if has("SAO PAULO") || has("BOVESPA") || has("B3") { return "🇧🇷" }
        if has("MEXICO") { return "🇲🇽" }
        if has("KOREA") || has("KOSPI") || has("KOSDAQ") { return "🇰🇷" }
        if has("TAIWAN") { return "🇹🇼" }
        if has("SINGAPORE") { return "🇸🇬" }
        if has("JAKARTA") || has("INDONESIA") { return "🇮🇩" }
        if has("PHILIPPINE") || has("MANILA") || e == "PSE" { return "🇵🇭" }
        if has("KUALA LUMPUR") || has("MALAYSIA") { return "🇲🇾" }
        if has("NASDAQ") || has("NYSE") || has("AMEX") || has("NYQ") || has("NMS")
            || has("PCX") || has("BATS") { return "🇺🇸" }
        return nil
    }

    private static func byCurrency(_ currency: String?) -> String? {
        switch currency?.uppercased() {
        case "CAD": return "🇨🇦"
        case "GBP", "GBP.L", "GBX": return "🇬🇧"
        case "EUR": return "🇪🇺"
        case "JPY": return "🇯🇵"
        case "AUD": return "🇦🇺"
        case "HKD": return "🇭🇰"
        case "CHF": return "🇨🇭"
        case "INR": return "🇮🇳"
        case "BRL": return "🇧🇷"
        case "SEK": return "🇸🇪"
        case "NOK": return "🇳🇴"
        case "DKK": return "🇩🇰"
        case "KRW": return "🇰🇷"
        case "TWD": return "🇹🇼"
        case "SGD": return "🇸🇬"
        case "MXN": return "🇲🇽"
        case "CNY": return "🇨🇳"
        case "NZD": return "🇳🇿"
        case "ISK": return "🇮🇸"
        case "PLN": return "🇵🇱"
        case "THB": return "🇹🇭"
        case "TRY": return "🇹🇷"
        case "ILS": return "🇮🇱"
        case "ZAR": return "🇿🇦"
        case "IDR": return "🇮🇩"
        case "PHP": return "🇵🇭"
        case "MYR": return "🇲🇾"
        case "USD": return "🇺🇸"
        default: return nil
        }
    }
}
